reframe/frontend/cli.py

```diff
@@ -183,6 +188,69 @@ def calc_verbosity(site_config, quiesce):
     return curr_verbosity - quiesce
 
 
+def distribute_tests(testcases, skip_system_check, skip_prgenv_check,
+                     node_map):
+    temporary_registry = None
+    new_checks = []
+    for t in testcases:
+        if not t.check.is_fixture():
+            cls = type(t.check)
+            basename = cls.__name__
+            original_var_info = cls.get_variant_info(
+                t.check.variant_num, recurse=True
+            )
+
+            def _rfm_distributed_set_run_nodes(obj):
+                if not obj.local:
+                    obj.job.pin_nodes = obj._rfm_nodelist
+
+            def _rfm_distributed_set_build_nodes(obj):
+                if not obj.local and not obj.build_locally:
+                    obj.build_job.pin_nodes = obj._rfm_nodelist
+
+            # We re-set the valid system and environment in a hook to
+            # make sure that it will not be overwriten by a parent
+            # post-init hook
+            def _rfm_distributed_set_valid_sys_env(obj):
+                obj.valid_systems = [t._partition.fullname]
+                obj.valid_prog_environs = [t._environ.name]
+
+            class BaseTest(t.check.__class__):
+                _rfm_nodelist = builtins.parameter(node_map[t._partition.fullname])
+                valid_systems = [t._partition.fullname]
+                valid_prog_environs = [t._environ.name]
+
+            nc = make_test(
+                f'__D_{t._partition.name}_{t._environ.name}_{basename}',
+                (BaseTest, ),
+                {},
+                methods=[
+                    builtins.run_before('run')(_rfm_distributed_set_run_nodes),
+                    builtins.run_before('compile')(_rfm_distributed_set_build_nodes),
+                    # TODO this hook is not working properly
+                    # builtins.run_after('init')(_rfm_distributed_set_valid_sys_env),
+                ]
+            )
+            # We have to set the prefix manually
+            nc._rfm_dynamic_test_prefix = t.check.prefix
+
+            for i in range(nc.num_variants):
+                # Check if this variant should be instantiated
+                var_info = copy.deepcopy(nc.get_variant_info(i, recurse=True))
+                var_info['params'].pop('_rfm_nodelist')
+                if var_info == original_var_info:
```

Comment:
@vkarak I finally remembered why I needed this... In case we have a parameterized test and we want to "distribute" one testcase of it with `-n TestFoo@1` we don't want to register all the variants, but only the ones that have already been selected.

Reply:
Still I would suggest to do it in a different way without having to rely on `get_variant_info()` or deep copies. I think you need to split the loop in two:

1. Loop over the test cases and collect the test classes and their variants in a dictionary. You can get the variant number of a test through the test's  `variant_num` attribute.
2. Loop over that dictionary and register only the variants that you have found.

Finally, instantiate all the tests in the registry.